File: structy/structy-py/2-Linked-List/create_linked_list.py
# Write a function, create_linked_list, that takes in a list of values as an argument. The function should create a linked list containing each item of the list as the values of the nodes. The function should return the head of the linked list.
#
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

#! Check answer
def create_linked_list(values):
    # first will need to start off creating a new head node
    head = Node(None)
    # return None if there are no values
    # will need to assign a tail for the following nodes / values
    tail = head
    # transverse through the length of the values list and assign to the linked list
    for i in range(len(values)):
        # tail.next should be assigned to the new values
        tail.next = Node(values[i])
        # assign tail to next
        tail = tail.next
    # return the head of the linked list at the end
    return head.next


#! TEST_00
create_linked_list(["h", "e", "y"])
# h -> e -> y

#! TEST_01
create_linked_list([1, 7, 1, 8])
# 1 -> 7 -> 1 -> 8

#! TEST_02
create_linked_list(["a"])
# a

#! TEST_03
create_linked_list([])
# null

File: structy/structy-js/Linked-List/createLinkedList.js
// Write a function, createLinkedList, that takes in an array of values as an argument.
// The function should create a linked list containing each element of the array as the values of the nodes.
// The function should return the head of the linked list.

class Node {
	constructor(val) {
		this.val = val;
		this.next = null;
	}
}

const createLinkedList = (values) => {
	// interate through array of values
	// set first value as new node
	// if there's no values return null

	let head = new Node(null);
	let tail = head;
	for (let i = 0; i < values.length; i++) {
		tail.next = new Node(values[i]);
		tail = tail.next;
	}
	// return head.next;
	return head;
};

//! recursive
// const createLinkedList = (values, i = 0) => {
//     if (i === values.length) return null;
//     const head = new Node(values[i]);
//     head.next = createLinkedList(values, i + 1);
//     return head;
//   };

//! TEST 00
console.log(createLinkedList(['h', 'e', 'y']));
// h -> e -> y

//! TEST 01
// console.log(createLinkedList([1, 7, 1, 8]));
// 1 -> 7 -> 1 -> 8
//! TEST 02
// console.log(createLinkedList(['a']));
// a
//! TEST 03
// console.log(createLinkedList([]));
// null
